GoogleForCreators/web-stories-wp
# Add custom Jest matchers for AMP validation

Adds the following new Jest matchers:

* `toBeValidAMP`
* `toBeValidAMPStoryPage` - like `toBeValidAMP`, but wraps the component/string in `<amp-story>` first
* `toBeValidAMPStoryElement` - like `toBeValidAMPStoryPage`, but wraps the component/string in `<amp-story-page>` and `<amp-story-grid-layer>` first

Usage:

```js
await expect(<VideoOutput {...props} />).toBeValidAMPStoryElement();
```

To test:

1. Apply the following patch:

```diff
diff --git assets/src/edit-story/elements/video/output.js assets/src/edit-story/elements/video/output.js
index d7d7c784..54255cb7 100644
--- assets/src/edit-story/elements/video/output.js
+++ assets/src/edit-story/elements/video/output.js
@@ -52,7 +52,7 @@ function VideoOutput({
     src,
   };
   const props = {
-    autoPlay: 'autoplay',
+    autoPlay: true,
     poster,
     layout: 'fill',
   };
```

2. Run `npm run test:js assets/src/edit-story/elements/video/test/output.js`
3. See the following output:

```
Video output › should produce valid AMP output

Expected <div style="position:absolute;width:177.77778%;height:100%;left:-38.88889%;top:0%"><amp-video autoPlay="true" poster="https://example.com/poster.png" layout="fill"><source type="video/mp4" src="https://example.com"/></amp-video></div> to be valid AMP. Errors:
The attribute 'autoplay' in tag 'amp-story >> amp-video' is set to the invalid value 'true'. (see https://amp.dev/documentation/components/amp-video)
```

This exact error was actually caught only because of the work on #346. I figured adding some tests for our AMP validity would make sense.

While it's not exactly how AMP validation works, and is not perfect (lots of assumptions right now), it seems to do the trick.

I also considered running the validator every time a story is saved, but `amphtml-validator` doesn't run in the browser.
